camera/src/Camera.cpp

```diff
@@ -0,0 +1,174 @@
+#include <rcbe/camera/Camera.hpp>
+
+#include <data_types/math/math_constants.hpp>
+
+namespace rcbe::rendering {
+Camera::Camera(const rendering::RenderingContextPtr &context, const math::Vector3d &position, const math::Vector3d &lookat, const math::Vector3d &up)
+:
+initial_position_ { position }
+, initial_lookat_ { lookat }
+, initial_up_ { up }
+, context_ { context }
+, transform_ { look_at(initial_position_, initial_lookat_, initial_up_) }
+{
+    if (!context_)
+        throw std::runtime_error("Rendering context for Camera is nullptr!");
+
+    BOOST_LOG_TRIVIAL(debug) << transform_.matrix();
+
+    update_context(transform_);
+}
+
+void Camera::update_context(const transform_type &t) const {
+    if (!context_)
+        throw std::runtime_error("Trying to access nullptr rendering context");
+
+    context_->update_transform(t);
+}
+
+math::Matrix4x4 Camera::look_at(const math::Vector3d& camera_position, const math::Vector3d& lookat, const math::Vector3d& up) {
+    position_ = camera_position;
+    direction_ = -((lookat - position_).normalized());  // direction is negated due to OpenGL's positive Z pointing towards camera,
+                                                        // and as such a vector pointing forward out of the camera is negative Z
+    right_ = math::Vector3d ::cross(up, direction_).normalized();
+    up_ = math::Vector3d ::cross(direction_, right_).normalized();
+
+    auto left_part = math::Matrix4x4 {
+            right_.x(), right_.y(), right_.z(), 0,
+            up_.x(), up_.y(), up_.z(), 0,
+            direction_.x(), direction_.y(), direction_.z(), 0,
+            0, 0, 0, 1
+    };
+
+    auto right_part = math::Matrix4x4 {
+            1, 0, 0, -position_.x(),
+            0, 1, 0, -position_.y(),
+            0, 0, 1, -position_.z(),
+            0, 0, 0, 1
+    };
+
+    return left_part * right_part;
+}
+
+const Camera::transform_type& Camera::get_transform() const {
+    std::lock_guard lg{translate_mutex_};
+    return transform_;
+}
+
+void Camera::reset_view() {
+    std::lock_guard lg{rotate_mutex_};
+    std::lock_guard lg2{translate_mutex_};
+    zoom_ = 45.;
+    context_->update_zoom(zoom_);
+    update_context(math::Transform(look_at(initial_position_, initial_lookat_, initial_up_)));
```

Comment:
How does this event work? `trnasform_` inside the `Camera` class is left unchanged. Figure it out.

Reply:
Fixed all of these through setting of transform inside `update_context`.